File: app/src/main/java/com/vitaliyhtc/dagger2investigation/presentation/productdetails/presenter/ProductDetailsPresenter.java
package com.vitaliyhtc.dagger2investigation.presentation.productdetails.presenter;

import com.vitaliyhtc.dagger2investigation.domain.ProductRepository;
import com.vitaliyhtc.dagger2investigation.domain.model.Product;
import com.vitaliyhtc.dagger2investigation.presentation.base.mvp.BasePresenter;
import com.vitaliyhtc.dagger2investigation.presentation.productdetails.view.ProductDetailsView;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class ProductDetailsPresenter implements BasePresenter<ProductDetailsView> {

    private ProductDetailsView mProductDetailsView;

    private ProductRepository mProductRepository;

    private CompositeDisposable mCompositeDisposable;

    public ProductDetailsPresenter(ProductRepository productRepository) {
        Timber.d("ProductDetailsPresenter: injected");
        mProductRepository = productRepository;
    }

    @Override
    public void onAttachView(ProductDetailsView view) {
        mProductDetailsView = view;
        mCompositeDisposable = new CompositeDisposable();
    }

    @Override
    public void onDetachView() {
        mProductDetailsView = null;

        if (mCompositeDisposable != null) {
            mCompositeDisposable.dispose();
        }
    }

    public void loadProduct(int productId) {
        Disposable disposable =
                mProductRepository.getOneProduct(productId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        product -> onProductLoaded(product),
                        this::loadProductsError
                );
        mCompositeDisposable.add(disposable);
    }

    private void onProductLoaded(Product product) {
        mProductDetailsView.showProduct(product);
    }

    private void loadProductsError(Throwable t) {
        mProductDetailsView.loadProductsError(t);
    }

}

File: app/src/main/java/com/vitaliyhtc/dagger2investigation/presentation/productdetails/presenter/ProductDetailsPresenter.kt
package com.vitaliyhtc.dagger2investigation.presentation.productdetails.presenter

import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.vitaliyhtc.dagger2investigation.domain.ProductRepository
import com.vitaliyhtc.dagger2investigation.domain.model.Product
import com.vitaliyhtc.dagger2investigation.presentation.productdetails.view.ProductDetailsView
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import timber.log.Timber

@InjectViewState
class ProductDetailsPresenter(private val productRepository: ProductRepository) : MvpPresenter<ProductDetailsView>() {

    private val mCompositeDisposable: CompositeDisposable = CompositeDisposable()
    private var isDataLoaded: Boolean = false

    private var mProduct: Product? = null

    override fun onDestroy() {
        super.onDestroy()
        mCompositeDisposable.dispose()
    }

    fun loadProduct(productId: Int) {
        if (isDataLoaded) return

        isDataLoaded = true

        productRepository.getOneProduct(productId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
//                .doOnSubscribe { d: Disposable -> mCompositeDisposable.add(d) }
                // for rxKotlin it can be even as { mCompositeDisposable += it }
                .doOnSubscribe { mCompositeDisposable.add(it) }
                .subscribe(
                        this::onProductLoaded,
                        this::loadProductsError
                )
    }

    private fun onProductLoaded(product: Product) {
        mProduct = product
        viewState.showProduct(product)
    }

    private fun loadProductsError(throwable: Throwable) {
        Timber.e(throwable)
        viewState.loadProductsError(throwable)
    }

    fun onFavoriteStatusChange(isFavorite: Boolean) {
        if (mProduct != null) {
            mProduct!!.is_favorite = isFavorite
            productRepository.updateProduct(mProduct!!)
        }
    }

}